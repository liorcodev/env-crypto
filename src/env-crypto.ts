import { createCipheriv, createDecipheriv, randomBytes, scryptSync } from 'crypto';
import fs from 'fs';
import path from 'path';

const ALGORITHM = 'aes-256-gcm';
export const DEFAULT_KEY = 'ENV_CRYPTO_KEY';

// Add simple error class
export class EnvCryptoError extends Error {
  constructor(
    message: string,
    public code: string,
  ) {
    super(message);
    this.name = 'EnvCryptoError';
  }
}

// Simple memory clearing helper
function clearBuffer(buffer: Buffer): void {
  if (buffer && Buffer.isBuffer(buffer)) {
    buffer.fill(0);
  }
}

/**
 * Validates and resolves a file path to ensure it is within the current working directory.
 * Prevents path traversal and absolute paths outside the project root.
 * @param filePath The file path to validate.
 * @returns The resolved absolute path.
 * @throws EnvCryptoError if the path is outside the working directory.
 */
function validatePath(filePath: string): string {
  const resolved = path.resolve(process.cwd(), filePath);
  const cwd = path.resolve(process.cwd());
  const relative = path.relative(cwd, resolved);

  // Prevent absolute paths outside CWD and path traversal
  if (relative.startsWith('..') || (path.isAbsolute(filePath) && !resolved.startsWith(cwd + path.sep))) {
    throw new EnvCryptoError('Invalid file path: outside working directory', 'INVALID_PATH');
  }
  return resolved;
}

/**
 * Encrypts a plaintext .env file and writes the encrypted output.
 * - Reads the source .env file.
 * - Uses AES-256-GCM with a random salt and IV.
 * - Stores salt, IV, encrypted content, and auth tag in output file.
 * @param sourcePath Path to the plaintext .env file.
 * @param outputPath Path to write the encrypted file.
 * @param keyEnvVar Name of the environment variable containing the encryption key.
 * @throws EnvCryptoError if the key is missing or file operations fail.
 */
export function encryptEnvFile(
  sourcePath: string = '.env',
  outputPath: string = '.env.encrypted',
  keyEnvVar: string = DEFAULT_KEY,
): void {
  let keyBuffer: Buffer | null = null;
  let salt: Buffer | null = null;

  try {
    // Get key from environment variable
    const key = process.env[keyEnvVar];
    if (!key) {
      throw new EnvCryptoError(`Environment variable ${keyEnvVar} not found`, 'KEY_NOT_FOUND');
    }

    // Validate and read the source file
    const validatedSourcePath = validatePath(sourcePath);
    const plaintext = fs.readFileSync(validatedSourcePath, 'utf8');

    // Generate random salt for each encryption
    salt = randomBytes(16);
    keyBuffer = scryptSync(key, salt, 32);

    // Generate initialization vector
    const iv = randomBytes(16);

    // Create cipher
    const cipher = createCipheriv(ALGORITHM, keyBuffer, iv);

    // Encrypt the plaintext
    let encrypted = cipher.update(plaintext, 'utf8', 'hex');
    encrypted += cipher.final('hex');

    // Get authentication tag for GCM mode
    const authTag = cipher.getAuthTag();

    // Write encrypted file with version, salt, IV and auth tag
    const encryptedData = JSON.stringify({
      version: '1.0.0',
      salt: salt.toString('hex'),
      iv: iv.toString('hex'),
      content: encrypted,
      authTag: authTag.toString('hex'),
    });

    const validatedOutputPath = validatePath(outputPath);
    fs.writeFileSync(validatedOutputPath, encryptedData, { mode: 0o600 });
    console.log(`Encrypted environment file saved to ${outputPath}`);
  } finally {
    // Clear sensitive data
    clearBuffer(keyBuffer!);
    clearBuffer(salt!);
  }
}

/**
 * Decrypts an encrypted .env file and returns the parsed environment variables.
 * - Reads and parses the encrypted file.
 * - Uses AES-256-GCM with salt and IV from the file.
 * - Returns a key-value object of environment variables.
 * @param sourcePath Path to the encrypted file.
 * @param keyEnvVar Name of the environment variable containing the decryption key.
 * @returns Record of environment variable key-value pairs.
 * @throws EnvCryptoError if the key is missing, file is invalid, or decryption fails.
 */
export function decryptEnvFile(
  sourcePath: string = '.env.encrypted',
  keyEnvVar: string = DEFAULT_KEY,
): Record<string, string> {
  let keyBuffer: Buffer | null = null;

  try {
    // Get key from environment variable
    const key = process.env[keyEnvVar];
    if (!key) {
      throw new EnvCryptoError(`Decryption key not found in environment variable ${keyEnvVar}`, 'KEY_NOT_FOUND');
    }

    // Validate and read encrypted file
    const validatedSourcePath = validatePath(sourcePath);
    const encryptedData = JSON.parse(fs.readFileSync(validatedSourcePath, 'utf8'));

    // Validate required fields exist
    if (!encryptedData.salt || !encryptedData.iv || !encryptedData.content || !encryptedData.authTag) {
      throw new EnvCryptoError('Invalid encrypted file format: missing required fields', 'INVALID_FORMAT');
    }

    // Prepare decryption with salt from file
    const salt = Buffer.from(encryptedData.salt, 'hex');
    keyBuffer = scryptSync(key, salt, 32);
    const iv = Buffer.from(encryptedData.iv, 'hex');
    const authTag = Buffer.from(encryptedData.authTag, 'hex');

    // Create decipher
    const decipher = createDecipheriv(ALGORITHM, keyBuffer, iv);
    decipher.setAuthTag(authTag);

    // Decrypt the content
    let decrypted = decipher.update(encryptedData.content, 'hex', 'utf8');
    decrypted += decipher.final('utf8');

    // Parse the decrypted content as .env format
    const envVars: Record<string, string> = {};
    decrypted.split('\n').forEach((line) => {
      // Skip comments and empty lines
      if (!line || line.startsWith('#')) return;

      // Parse key-value pairs (supports quoted values)
      const match = line.match(/^\s*([\w.-]+)\s*=\s*(.*)?\s*$/);
      if (match) {
        const key = match[1];
        let value = match[2] || '';

        // Remove quotes if present
        if (
          value.length > 1 &&
          ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'")))
        ) {
          value = value.substring(1, value.length - 1);
        }

        if (typeof key !== 'undefined') {
          envVars[key] = value;
        }
      }
    });

    return envVars;
  } finally {
    // Clear sensitive data
    clearBuffer(keyBuffer!);
  }
}
