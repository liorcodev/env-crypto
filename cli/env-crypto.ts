#!/usr/bin/env node
import { encryptEnvFile, decryptEnvFile } from '../src/env-crypto';
import fs from 'fs';
import path from 'path';

const args = process.argv.slice(2);
const command = args[0];
const DEFAULT_KEY = 'ENV_CRYPTO_KEY';
if (command === 'encrypt') {
  const sourcePath = args[1] || '.env';
  const outputPath = args[2] || '.env.encrypted';
  const keyEnvVar = args[3] || DEFAULT_KEY;
  encryptEnvFile(sourcePath, outputPath, keyEnvVar);
} else if (command === 'decrypt') {
  const sourcePath = args[1] || '.env.encrypted';
  const outputPath = args[2] || '.env';
  const keyEnvVar = args[3] || DEFAULT_KEY;

  const vars = decryptEnvFile(sourcePath, keyEnvVar);
  console.log('Decrypted variables:', Object.keys(vars));

  // Write back to .env file if output path is provided
  if (outputPath) {
    let envContent = '';
    // Format the variables as key=value pairs
    Object.entries(vars).forEach(([key, value]) => {
      // Check if value needs quotes (if it contains spaces or special characters)
      if (/[\s,;]/.test(value)) {
        envContent += `${key}="${value}"\n`;
      } else {
        envContent += `${key}=${value}\n`;
      }
    });

    fs.writeFileSync(path.resolve(process.cwd(), outputPath), envContent);
    console.log(`Decrypted environment saved to ${outputPath}`);
  }
} else {
  console.log(`
Usage:
  env-crypto encrypt [sourcePath] [outputPath] [keyEnvVar]
  env-crypto decrypt [sourcePath] [outputPath] [keyEnvVar]

You can also use npx:
  npx env-crypto encrypt [sourcePath] [outputPath] [keyEnvVar]
  npx env-crypto decrypt [sourcePath] [outputPath] [keyEnvVar]
  `);
}
