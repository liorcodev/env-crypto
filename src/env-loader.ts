import { decryptEnvFile, DEFAULT_KEY } from './env-crypto';

/**
 * Processes a single environment variable value.
 * - Converts numeric strings to numbers.
 * - Converts comma-separated strings to string arrays.
 * - Returns plain strings otherwise.
 */
const processEnvValue = (value: string): string | number | string[] => {
  // Check if it's a number (integer or decimal)
  if (/^-?\d+(\.\d+)?$/.test(value)) {
    return Number(value);
  }

  // Check if it's a comma-separated array
  if (value.includes(',')) {
    return value.split(',').map((item) => item.trim());
  }

  // Default: return as string
  return value;
};

// The main environment object, extended from process.env
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const env: { [key: string]: any } & typeof process.env = {} as typeof process.env;
let isInitialized = false;

/**
 * Initializes environment variables from the encrypted file.
 * - Loads and decrypts the environment file.
 * - Processes each variable (number, array, string).
 * - Populates the `env` object and sets process.env for arrays as JSON strings.
 * - Only runs once per process unless reset.
 * @returns The populated env object.
 */
function initEnv(sourcePath: string = '.env.encrypted', keyEnvVar: string = DEFAULT_KEY): typeof env {
  if (isInitialized) {
    return env;
  }

  try {
    // Load and decrypt env vars
    const envVars = decryptEnvFile(sourcePath, keyEnvVar);

    // Iterate over the decrypted environment variables
    Object.entries(envVars).forEach(([key, value]) => {
      if (!env[key]) {
        const processedValue = processEnvValue(value);

        // Assign the processed value to the env object
        if (typeof processedValue === 'string') {
          env[key] = processedValue;
        } else if (typeof processedValue === 'number') {
          env[key] = processedValue;
        } else if (Array.isArray(processedValue)) {
          env[key] = processedValue;
          // For process.env compatibility, convert arrays to JSON strings
          process.env[key] = JSON.stringify(processedValue);
        }
      }
    });

    isInitialized = true;
    console.log('Decrypted environment variables loaded');
    return env;
  } catch (error) {
    // Wrap and rethrow errors with context
    throw new Error('Failed to load encrypted environment variables', { cause: error });
  }
}

/**
 * Resets the initialization state and clears the env object.
 * Useful for testing or reloading environment variables.
 */
function resetEnv() {
  isInitialized = false;
  Object.keys(env).forEach((key) => delete env[key]);
}

export { env, initEnv, resetEnv };
