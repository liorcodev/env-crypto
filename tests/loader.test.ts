/**
 * Tests for env-loader module.
 * Ensures environment variables are loaded, parsed, and initialized correctly from encrypted files.
 */
import { describe, it, expect, afterAll } from 'bun:test';
import { encryptEnvFile } from '../src/env-crypto';
import { env, initEnv, resetEnv } from '../src/env-loader';

describe('Environment Loader', () => {
  const ENV_CRYPTO_KEY = 'mkNA802Hqwxpl6c0';

  // Clean up any test files created during the tests
  afterAll(async () => {
    const envFile = Bun.file('.env');
    const encryptedEnvFile = Bun.file('.env.encrypted');
    if (await envFile.exists()) envFile.delete();
    if (await encryptedEnvFile.exists()) encryptedEnvFile.delete();
  });

  it('should initialize environment variables', async () => {
    // Arrange: Write a test .env file and encrypt it
    const sourcePath = '.env';
    const outputPath = '.env.encrypted';

    process.env['ENV_CRYPTO_KEY'] = ENV_CRYPTO_KEY;
    await Bun.write(sourcePath, 'TEST_VAR=123\nANOTHER_VAR=hello,world\nEXTRA_VAR=extra\n');

    // Encrypt the environment file
    encryptEnvFile(sourcePath, outputPath);

    // Wait briefly to ensure file is written
    await new Promise((resolve) => setTimeout(resolve, 10));

    // Act: Initialize environment variables from encrypted file
    initEnv();

    // Assert: Check parsed values
    expect(env.TEST_VAR).toBe(123);
    expect(env.ANOTHER_VAR).toEqual(['hello', 'world']);
    expect(env.EXTRA_VAR).toBe('extra');
  });

  it('should execute only once even if called multiple times', async () => {
    // Arrange: Write and encrypt a test .env file
    const sourcePath = '.env';
    const outputPath = '.env.encrypted';

    process.env['ENV_CRYPTO_KEY'] = ENV_CRYPTO_KEY;
    await Bun.write(sourcePath, 'TEST_VAR=123');

    // Encrypt the environment file
    encryptEnvFile(sourcePath, outputPath);

    // Wait briefly to ensure file is written
    await new Promise((resolve) => setTimeout(resolve, 10));

    // Act: Call initEnv multiple times
    initEnv();
    const returnEnv = initEnv(); // Should return the same object

    // Assert: Check that the returned object is the same and values are correct
    expect(returnEnv).toBe(env);
    expect(returnEnv.TEST_VAR).toBe(123);
  });

  it('should throw error when decryption fails', async () => {
    // Arrange: Reset the initialization state and remove the encryption key
    resetEnv();
    delete process.env['ENV_CRYPTO_KEY'];

    // Act & Assert: Attempt to initialize environment and expect an error
    expect(() => {
      initEnv();
    }).toThrowErrorMatchingInlineSnapshot(`"Failed to load encrypted environment variables"`);

    // Restore the key for other tests
    process.env['ENV_CRYPTO_KEY'] = ENV_CRYPTO_KEY;
  });
});
