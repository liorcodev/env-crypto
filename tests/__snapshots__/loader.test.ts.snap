// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Environment Loader should throw error when decryption fails 1`] = `"Failed to load encrypted environment variables"`;
